File: src/ai/agents/heuristic_agent.cpp
#include "heuristic_agent.hpp"
#include <algorithm>
#include <iostream>

namespace dm::ai {

    HeuristicAgent::HeuristicAgent(int player_id, const std::map<dm::core::CardID, dm::core::CardDefinition>& card_db)
        : player_id_(player_id), card_db_(card_db) {
        // Initialize RNG with a random seed
        std::random_device rd;
        rng_ = std::mt19937(rd());
    }

    const dm::core::CardDefinition* HeuristicAgent::get_def(dm::core::CardID cid) const {
        auto it = card_db_.find(cid);
        if (it != card_db_.end()) {
            return &it->second;
        }
        return nullptr;
    }

    dm::core::CardID HeuristicAgent::get_card_id(const dm::core::GameState& state, int instance_id) const {
        // Safe access (const method assumed on GameState)
        // If get_card_instance is not const, we might need const_cast or rely on API
        // Assuming const auto* get_card_instance(int) const exists.
        // If not, we might have issues. Let's assume standard const access.
        // Actually, state.players is public, we can scan. But get_card_instance is better.
        // Based on other files, it seems to return pointer.
        // The implementation in GameState likely supports const.
        // If compilation fails, we fix it.
        const auto* ptr = state.get_card_instance(instance_id);
        if (ptr) return ptr->card_id;
        return 0;
    }

    dm::core::CommandDef HeuristicAgent::get_command(const dm::core::GameState& state,
                                                const std::vector<dm::core::CommandDef>& legal_actions) {
        if (legal_actions.empty()) {
            return dm::core::CommandDef(); // Should not happen if game is not over
        }

        using namespace dm::core;

        // 1. Mana Charge
        std::vector<CommandDef> mana_actions;
        for (const auto& action : legal_actions) {
            if (action.type == CommandType::MANA_CHARGE ||
               (action.type == CommandType::MOVE_CARD && action.to_zone == "MANA_ZONE")) {
                mana_actions.push_back(action);
            }
        }

        if (!mana_actions.empty()) {
            int current_mana = state.players[player_id_].mana_zone.size();
            if (current_mana < 8) {
                // Return random mana charge
                std::uniform_int_distribution<> dist(0, mana_actions.size() - 1);
                return mana_actions[dist(rng_)];
            }
        }

        // 2. Play Card
        std::vector<CommandDef> play_actions;
        for (const auto& action : legal_actions) {
            if (action.type == CommandType::PLAY_FROM_ZONE ||
                action.type == CommandType::CAST_SPELL ||
                action.type == CommandType::SUMMON_TOKEN) {
                play_actions.push_back(action);
            }
        }

        if (!play_actions.empty()) {
            // Play the most expensive card possible
            std::sort(play_actions.begin(), play_actions.end(), [&](const CommandDef& a, const CommandDef& b) {
                CardID cid_a = get_card_id(state, a.instance_id);
                CardID cid_b = get_card_id(state, b.instance_id);
                const auto* def_a = get_def(cid_a);
                const auto* def_b = get_def(cid_b);
                int cost_a = def_a ? def_a->cost : 0;
                int cost_b = def_b ? def_b->cost : 0;
                return cost_a > cost_b; // Descending
            });
            return play_actions[0];
        }

        // 3. Attack Player (Aggro)
        std::vector<CommandDef> attack_player_actions;
        for (const auto& action : legal_actions) {
            if (action.type == CommandType::ATTACK_PLAYER) {
                attack_player_actions.push_back(action);
            }
        }

        if (!attack_player_actions.empty()) {
            std::uniform_int_distribution<> dist(0, attack_player_actions.size() - 1);
            return attack_player_actions[dist(rng_)];
        }

        // 4. Attack Creature
        std::vector<CommandDef> attack_creature_actions;
        for (const auto& action : legal_actions) {
            if (action.type == CommandType::ATTACK_CREATURE) {
                attack_creature_actions.push_back(action);
            }
        }

        if (!attack_creature_actions.empty()) {
            std::uniform_int_distribution<> dist(0, attack_creature_actions.size() - 1);
            return attack_creature_actions[dist(rng_)];
        }

        // 5. Block
        std::vector<CommandDef> block_actions;
        for (const auto& action : legal_actions) {
            if (action.type == CommandType::BLOCK) {
                block_actions.push_back(action);
            }
        }

        if (!block_actions.empty()) {
            // Block if shields are low or 50% chance
            int shield_count = state.players[player_id_].shield_zone.size();
            bool should_block = false;
            if (shield_count <= 2) {
                should_block = true;
            } else {
                std::uniform_real_distribution<> dist(0.0, 1.0);
                if (dist(rng_) < 0.5) {
                    should_block = true;
                }
            }

            if (should_block) {
                std::uniform_int_distribution<> dist(0, block_actions.size() - 1);
                return block_actions[dist(rng_)];
            }
        }

        // 6. Select Target
        std::vector<CommandDef> select_actions;
        for (const auto& action : legal_actions) {
            if (action.type == CommandType::SELECT_TARGET) {
                select_actions.push_back(action);
            }
        }

        if (!select_actions.empty()) {
            std::uniform_int_distribution<> dist(0, select_actions.size() - 1);
            return select_actions[dist(rng_)];
        }

        // 7. Shield Trigger
        std::vector<CommandDef> st_actions;
        for (const auto& action : legal_actions) {
            if (action.type == CommandType::SHIELD_TRIGGER) {
                st_actions.push_back(action);
            }
        }

        if (!st_actions.empty()) {
            // Always use first trigger for now
            return st_actions[0];
        }

        // Default: Random choice from all legal actions
        std::uniform_int_distribution<> dist(0, legal_actions.size() - 1);
        return legal_actions[dist(rng_)];
    }

}

File: dm_toolkit/ai/agent/heuristic_agent.py

from typing import Any, List, Dict, Optional


class HeuristicAgent:
    def __init__(self, player_id: int) -> None:
        self.player_id: int = player_id

    def get_command(self, state: Any, legal_commands: List[Any], card_db: Dict[int, Any]) -> Optional[Any]:
        import dm_ai_module
        import random

        # 再発防止: dm_ai_module.CommandType を使用。ActionType は C++ レガシースタブ。
        CommandType = getattr(dm_ai_module, 'CommandType', None) or getattr(dm_ai_module, 'ActionType', None)

        def _type_is(cmd_obj: Any, name: str) -> bool:
            try:
                t = getattr(cmd_obj, 'type', None)
                if CommandType is not None and hasattr(CommandType, name):
                    return bool(t == getattr(CommandType, name))
                # Fallbacks: compare by enum name or string match
                if isinstance(t, str) and t.upper().endswith(name):
                    return True
                try:
                    return bool(getattr(t, 'name', '').upper() == name)
                except Exception:
                    return False
            except Exception:
                return False

        if not legal_commands:
            return None

        def get_def(cid: int) -> Optional[Any]:
            if cid in card_db:
                return card_db[cid]
            return None

        # 1. Mana Charge
        mana_cmds = [c for c in legal_commands if _type_is(c, 'MANA_CHARGE')]
        if mana_cmds:
            current_mana = len(state.players[self.player_id].mana_zone)
            if current_mana < 8:
                return random.choice(mana_cmds)

        # 2. Play Card
        play_cmds = [c for c in legal_commands if _type_is(c, 'PLAY_CARD') or _type_is(c, 'DECLARE_PLAY') or _type_is(c, 'PLAY_FROM_ZONE')]
        if play_cmds:
            def get_cost(cmd: Any) -> int:
                cdef = get_def(getattr(cmd, 'card_id', 0))
                return int(cdef.cost) if cdef and hasattr(cdef, 'cost') else 0

            play_cmds.sort(key=get_cost, reverse=True)
            return play_cmds[0]

        # 3. Attack Player (Aggro)
        attack_player_cmds = [c for c in legal_commands if _type_is(c, 'ATTACK_PLAYER')]
        if attack_player_cmds:
            return random.choice(attack_player_cmds)

        # 4. Attack Creature
        attack_creature_cmds = [c for c in legal_commands if _type_is(c, 'ATTACK_CREATURE')]
        if attack_creature_cmds:
            return random.choice(attack_creature_cmds)

        # 5. Block
        block_cmds = [c for c in legal_commands if _type_is(c, 'BLOCK')]
        if block_cmds:
            if len(state.players[self.player_id].shield_zone) <= 2:
                return random.choice(block_cmds)
            if random.random() < 0.5:
                return random.choice(block_cmds)

        # 6. Select Target (for effects)
        select_cmds = [c for c in legal_commands if _type_is(c, 'SELECT_TARGET')]
        if select_cmds:
            return random.choice(select_cmds)

        # 7. Shield Trigger
        st_cmds = [c for c in legal_commands if _type_is(c, 'USE_SHIELD_TRIGGER')]
        if st_cmds:
            return st_cmds[0]

        return random.choice(legal_commands)

    # 後方互換エイリアス。新規コードでは get_command を使用すること。
    get_action = get_command
